From a91fe615b662cf809d6bdd44a6dc6b43208133a2 Mon Sep 17 00:00:00 2001
From: Harrison Booth <harrisontbooth@gmail.com>
Date: Sat, 17 Oct 2026 09:14:27 -0500
Subject: [PATCH 2/2] Keep external include flags after PostgreSQL headers and
 in fallback probes

Focused follow-up after Aidar Imamov CF7300 v4. Based on Andres Freund extra_include_dirs prototype and Thomas Munro include-order discussion. Retains author v4 export implementation; excludes competing exporter, ICU discovery and LLVM extension-interface changes. No runtime claim from packaging.
---
 meson.build | 30 +++++++++++++++++++++++-------
 1 file changed, 23 insertions(+), 7 deletions(-)

diff --git a/meson.build b/meson.build
index 6bf91cc..d6b0036 100644
--- a/meson.build
+++ b/meson.build
@@ -115,7 +115,6 @@ endif
 ###############################################################
 
 postgres_inc_d = ['src/include']
-postgres_inc_d += get_option('extra_include_dirs')
 
 postgres_lib_d = get_option('extra_lib_dirs')
 
@@ -390,6 +389,23 @@ else
   error('unknown host system: @0@'.format(host_system))
 endif
 
+# While the most obvious way to implement extra_include_dirs would be to add
+# extra_include_dirs to postgres_inc_d, that turns out to not work well: For
+# some parts of the buildtree (e.g. src/interfaces/libpq or tools linking
+# against libpq) we need to add additional directories to be included, and
+# those directories need to take precedence over extra_include_dirs.  The
+# easiest way to achieve that is to gin up the compiler arguments ourselves
+# (which luckily is the same across all supported compilers) and add it to
+# cppflags. That works as meson always adds explicit include_directories
+# before generic compiler arguments.
+#
+# An additional advantage is that this way the extra include directories are
+# included in pg_config --cppflags output, keeping the behavior consistent
+# with autoconf builds.
+foreach extra_dir : get_option('extra_include_dirs')
+  cppflags += '-I@0@'.format(extra_dir)
+endforeach
+
 
 
 ###############################################################
@@ -856,7 +872,7 @@ else
   # installed
   if not ldap.found()
     ldap = cc.find_library('ldap', required: ldapopt, dirs: test_lib_d,
-      has_headers: 'ldap.h', header_include_directories: postgres_inc)
+      has_headers: 'ldap.h', header_args: test_c_args, header_include_directories: postgres_inc)
 
     # The separate ldap_r library only exists in OpenLDAP < 2.5, and if we
     # have 2.5 or later, we shouldn't even probe for ldap_r (we might find a
@@ -871,7 +887,7 @@ else
 
       # Use ldap_r for FE if available, else assume ldap is thread-safe.
       ldap_r = cc.find_library('ldap_r', required: false, dirs: test_lib_d,
-        has_headers: 'ldap.h', header_include_directories: postgres_inc)
+        has_headers: 'ldap.h', header_args: test_c_args, header_include_directories: postgres_inc)
       if not ldap_r.found()
         ldap_r = ldap
       else
@@ -1644,7 +1660,7 @@ if sslopt in ['auto', 'openssl']
   if not ssl.found()
     ssl_lib = cc.find_library('ssl',
       dirs: test_lib_d,
-      header_include_directories: postgres_inc,
+      header_args: test_c_args, header_include_directories: postgres_inc,
       has_headers: ['openssl/ssl.h', 'openssl/err.h'],
       required: openssl_required)
     crypto_lib = cc.find_library('crypto',
@@ -1746,12 +1762,12 @@ if uuidopt != 'none'
     if not uuid.found()
       uuid = cc.find_library('ossp-uuid',
         required: false, dirs: test_lib_d,
-        has_headers: uuidheader, header_include_directories: postgres_inc)
+        has_headers: uuidheader, header_args: test_c_args, header_include_directories: postgres_inc)
     endif
     if not uuid.found()
       uuid = cc.find_library('uuid',
         required: true, dirs: test_lib_d,
-        has_headers: uuidheader, header_include_directories: postgres_inc)
+        has_headers: uuidheader, header_args: test_c_args, header_include_directories: postgres_inc)
     endif
   else
     error('unknown uuid build option value: @0@'.format(uuidopt))
@@ -3333,7 +3349,7 @@ if not nlsopt.disabled()
     else
       libintl = cc.find_library('intl',
         has_headers: ['libintl.h'], required: nlsopt,
-        header_include_directories: postgres_inc,
+        header_args: test_c_args, header_include_directories: postgres_inc,
         dirs: test_lib_d)
     endif
   endif
-- 
2.53.0.windows.2

